# Pandas practice:

#### Import necessary modules:

Model - random forest;

In [1]:
import numpy as np
import pandas as pd

from sklearn.ensemble import RandomForestClassifier

#### Create train and test dataframes:

In [2]:
train_dataframe = pd.read_csv("train.csv")
test_dataframe = pd.read_csv("test.csv")

#### Check train and test dataframe:

In [3]:
train_dataframe.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
test_dataframe.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

They structures are similar.

#### Analyzing train dataframe:

In [5]:
train_dataframe.shape

(891, 12)

In [6]:
train_dataframe.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [7]:
train_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


We have nan values, we need to delete this rows or put something in them.

#### Prepare train dataframe:

In [8]:
feature_columns = ["Pclass", "Sex", "Age"]
predict_column = ["Survived"]

In [9]:
samples = train_dataframe[feature_columns]

#### Check our features dataframe:

In [10]:
samples.head()

Pclass     Sex   Age
0       3    male  22.0
1       1  female  38.0
2       3  female  26.0
3       1  female  35.0
4       3    male  35.0

In [11]:
samples.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Pclass  891 non-null    int64  
 1   Sex     891 non-null    object 
 2   Age     714 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 21.0+ KB


We have nan values, if we drop them, than we loss over (891 - 714 = 177) 20 % of samples. By other side we can fill them by median or mode.

The way to drop nan values:

In [12]:
# samples = samples.dropna()

The way with median:

In [13]:
# age_median = train_dataframe["Age"].median()

# age_median

In [14]:
# samples["Age"].fillna(age_median, inplace=True)

The way with mean:

In [15]:
age_mean = train_dataframe["Age"].mean()

age_mean

29.69911764705882

In [16]:
samples["Age"].fillna(age_mean, inplace=True)

C:\Users\Deniz\AppData\Local\Temp\ipykernel_16404\3760982822.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  samples["Age"].fillna(age_mean, inplace=True)


Change labels:

In [17]:
samples = samples.rename(columns={"Pclass": "Class"})

In [18]:
samples.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Class   891 non-null    int64  
 1   Sex     891 non-null    object 
 2   Age     891 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 21.0+ KB


#### Check prepare data:

In [19]:
samples.head()

Class     Sex   Age
0      3    male  22.0
1      1  female  38.0
2      3  female  26.0
3      1  female  35.0
4      3    male  35.0

#### Prepare data for fit:

Convert data from categorie feature to bool:

In [20]:
X = pd.get_dummies(samples)

Check what we have:

In [21]:
X.head()

Class   Age  Sex_female  Sex_male
0      3  22.0           0         1
1      1  38.0           1         0
2      3  26.0           1         0
3      1  35.0           1         0
4      3  35.0           0         1

The same for testing data:

In [22]:
X_test = pd.get_dummies(test_dataframe[feature_columns])

Check the result:

In [23]:
X_test.head()

Pclass   Age  Sex_female  Sex_male
0       3  34.5           0         1
1       3  47.0           1         0
2       2  62.0           0         1
3       3  27.0           0         1
4       3  22.0           1         0

In [24]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Pclass      418 non-null    int64  
 1   Age         332 non-null    float64
 2   Sex_female  418 non-null    uint8  
 3   Sex_male    418 non-null    uint8  
dtypes: float64(1), int64(1), uint8(2)
memory usage: 7.5 KB


Fill empty cells in test data:

In [25]:
X_test["Age"].fillna(age_mean, inplace=True)

In [26]:
X_test.rename(columns={"Pclass": "Class"}, inplace=True)

In [27]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Class       418 non-null    int64  
 1   Age         418 non-null    float64
 2   Sex_female  418 non-null    uint8  
 3   Sex_male    418 non-null    uint8  
dtypes: float64(1), int64(1), uint8(2)
memory usage: 7.5 KB


Prepare targets:

In [28]:
Y = train_dataframe.Survived

Check targets:

In [29]:
Y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 891 entries, 0 to 890
Series name: Survived
Non-Null Count  Dtype
--------------  -----
891 non-null    int64
dtypes: int64(1)
memory usage: 7.1 KB


#### Create model:

In [30]:
model = RandomForestClassifier(n_estimators=100, max_depth=4, random_state=1)

#### Fit the model:

In [31]:
model.fit(X, Y)

RandomForestClassifier(max_depth=4, random_state=1)

In [32]:
predictions = model.predict(X_test)

Concatenate prediction:

In [33]:
X_test = pd.concat([X_test, pd.Series(predictions)], axis=1)

Prepare final data:

In [34]:
X_test.rename(columns={0: "Survived"}, inplace=True)

Check it:

In [35]:
X_test.head()

Class   Age  Sex_female  Sex_male  Survived
0      3  34.5           0         1         0
1      3  47.0           1         0         0
2      2  62.0           0         1         0
3      3  27.0           0         1         0
4      3  22.0           1         0         1

Create result dataframe:

In [36]:
output = pd.DataFrame({"PassengerId": test_dataframe.PassengerId, "Survived": predictions})

Check result:

In [37]:
output.head()

PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         1

Save result to csv:

In [38]:
output.to_csv("submission.csv", index=False)